File: jscomp/others/bs_HashMapString.ml
# 2 "hashmap.cppo.ml"
type key = string
type seed = int
external caml_hash_mix_string : seed -> string -> seed  = "caml_hash_mix_string"
external final_mix : seed -> seed = "caml_hash_final_mix"
let hash (s : key) =   
  final_mix  (caml_hash_mix_string 0 s )

# 19
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)
(**  Adapted by Authors of BuckleScript 2017                           *)
module N = Bs_internalBuckets
module C = Bs_internalBucketsType
module A = Bs_Array
type ('a, 'b,'id) t0 = ('a,'b) N.t0 

type 'b t = (key,'b,unit) t0


let rec insert_bucket  ~h_buckets ~ndata_tail h old_bucket = 
  match C.toOpt old_bucket with 
  | None -> ()
  | Some cell ->
    let nidx = hash (N.key cell) land (Array.length h_buckets - 1) in 
    let v = C.return cell in 
    begin match C.toOpt (Bs_Array.unsafe_get ndata_tail nidx) with
      | None -> 
        Bs_Array.unsafe_set h_buckets nidx  v
      | Some tail ->
        N.nextSet tail v  (* cell put at the end *)            
    end;          
    Bs_Array.unsafe_set ndata_tail nidx  v;
    insert_bucket  ~h_buckets ~ndata_tail h (N.next cell)


let resize  h =
  let odata = C.buckets h in
  let osize = Array.length odata in
  let nsize = osize * 2 in
  if nsize >= osize then begin (* no overflow *)
    let h_buckets = A.makeUninitialized nsize  in
    let ndata_tail = A.makeUninitialized nsize  in (* keep track of tail *)
    C.bucketsSet h  h_buckets;          (* so that indexfun sees the new bucket count *)
    for i = 0 to osize - 1 do
      insert_bucket  ~h_buckets ~ndata_tail h (Bs_Array.unsafe_get odata i)
    done;
    for i = 0 to nsize - 1 do
      match C.toOpt (Bs_Array.unsafe_get ndata_tail i) with
      | None -> ()
      | Some tail -> N.nextSet tail C.emptyOpt
    done
  end


let rec replace_bucket  (key : key) info cell = 
    if  (N.key cell) = key 
    then
      begin
        N.keySet cell key;
        N.valueSet cell info;
        false
      end
    else
    match C.toOpt (N.next cell) with 
    | None -> true 
    | Some cell -> 
      replace_bucket key info cell

let set  h (key : key) value =
  let h_buckets = C.buckets h in 
  let i = hash key land (Array.length h_buckets - 1) in 
  let l = Array.unsafe_get h_buckets i in  
  match C.toOpt l with 
  | None -> 
    Bs_Array.unsafe_set h_buckets i (C.return (N.bucket ~key ~value ~next:l));
    C.sizeSet h (C.size h + 1);
    if C.size h > Array.length (C.buckets h) lsl 1 then resize h 
  | Some bucket -> 
    begin 
    if replace_bucket key value bucket then begin
      Bs_Array.unsafe_set h_buckets i (C.return 
                                       (N.bucket ~key ~value ~next:l));
      C.sizeSet h (C.size h + 1);
      if C.size h > Array.length (C.buckets h) lsl 1 then resize  h
    end   
    end

let rec remove_bucket h h_buckets  i (key : key) prec buckets =
  match C.toOpt buckets with
  | None -> ()
  | Some cell  ->
    let cell_next = N.next cell in 
    if  N.key cell = key 
    then 
      begin
        N.nextSet prec cell_next;
        C.sizeSet h (C.size h - 1);        
      end
    else remove_bucket  h h_buckets i key cell cell_next

let remove  h key =  
  let h_buckets = C.buckets h in 
  let i = hash key land (Array.length h_buckets - 1) in  
  let bucket = (Bs_Array.unsafe_get h_buckets i) in 
  match C.toOpt bucket with 
  | None -> ()
  | Some cell -> 
    if (N.key cell) = key then 
    begin 
      Bs_Array.unsafe_set h_buckets i (N.next cell);
      C.sizeSet h (C.size h - 1)
    end 
    else 
      remove_bucket  h h_buckets i key cell (N.next cell)



let rec findAux  (key : key) buckets = 
  match C.toOpt buckets with 
  | None ->
    None
  | Some cell ->
    if key = (N.key cell)  then Some (N.value cell) 
    else findAux key  (N.next cell)

let get  h (key : key) =
  let h_buckets = C.buckets h in 
  let nid = hash key  land (Array.length h_buckets - 1) in 
  match C.toOpt @@ Bs_Array.unsafe_get h_buckets nid with
  | None -> None
  | Some cell1 ->
    if key = (N.key cell1)  then Some (N.value cell1) else
      match C.toOpt (N.next cell1) with
      | None -> None
      | Some cell2 ->
        if  key = (N.key cell2)  then Some (N.value cell2) else
          match C.toOpt (N.next cell2) with
          | None -> None
          | Some cell3 ->
            if  key = (N.key cell3)  then Some (N.value cell3)
            else findAux  key (N.next cell3)




let rec mem_in_bucket (key : key) cell = 
    (N.key cell)  = key  ||
    (match C.toOpt (N.next cell) with 
    | None -> false
    | Some nextCell -> mem_in_bucket  key nextCell
     )
    
let has  h key =
  let h_buckets = C.buckets h in 
  let nid = hash key land (Array.length h_buckets - 1) in 
  let bucket = Bs_Array.unsafe_get h_buckets nid in 
  match C.toOpt bucket with 
  | None -> false
  | Some bucket -> 
    mem_in_bucket  key bucket


let create = C.create0
let clear = C.clear0
let reset = C.reset0
let size = C.length0
let forEach = N.iter0
let reduce = N.fold0
let logStats = N.logStats0
let filterMapDone = N.filterMapInplace0
let filterMap h f = filterMapDone h f; h
let toArray = N.toArray0 

let ofArray arr  = 
  let len = Bs.Array.length arr in 
  let v = create len in 
  for i = 0 to len - 1 do 
    let k,value = (Bs.Array.unsafe_get arr i) in
    set v k value
  done ;
  v

(* TOOD: optimize heuristics for resizing *)  
let mergeArrayDone h arr =   
  let len = Bs.Array.length arr in 
  for i = 0 to len - 1 do 
    let k,v = (Bs_Array.unsafe_get arr i) in
    set h k v 
done

let mergeArray h arr = mergeArrayDone h arr; h

File: lib/js/bs_HashMapString.js
'use strict';

var Caml_hash = require("./caml_hash.js");
var Bs_internalBuckets = require("./bs_internalBuckets.js");
var Bs_internalBucketsType = require("./bs_internalBucketsType.js");

function hash(s) {
  return Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_string(0, s));
}

function copyBucketReHash(h_buckets, ndata_tail, _old_bucket) {
  while(true) {
    var old_bucket = _old_bucket;
    if (old_bucket !== undefined) {
      var s = old_bucket.key;
      var nidx = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_string(0, s)) & (h_buckets.length - 1 | 0);
      var match = ndata_tail[nidx];
      if (match !== undefined) {
        match.next = old_bucket;
      } else {
        h_buckets[nidx] = old_bucket;
      }
      ndata_tail[nidx] = old_bucket;
      _old_bucket = old_bucket.next;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function resize(h) {
  var odata = h.buckets;
  var osize = odata.length;
  var nsize = (osize << 1);
  if (nsize >= osize) {
    var h_buckets = new Array(nsize);
    var ndata_tail = new Array(nsize);
    h.buckets = h_buckets;
    for(var i = 0 ,i_finish = osize - 1 | 0; i <= i_finish; ++i){
      copyBucketReHash(h_buckets, ndata_tail, odata[i]);
    }
    for(var i$1 = 0 ,i_finish$1 = nsize - 1 | 0; i$1 <= i_finish$1; ++i$1){
      var match = ndata_tail[i$1];
      if (match !== undefined) {
        match.next = Bs_internalBucketsType.emptyOpt;
      }
      
    }
    return /* () */0;
  } else {
    return 0;
  }
}

function replaceInBucket(key, info, _cell) {
  while(true) {
    var cell = _cell;
    if (cell.key === key) {
      cell.value = info;
      return /* false */0;
    } else {
      var match = cell.next;
      if (match !== undefined) {
        _cell = match;
        continue ;
        
      } else {
        return /* true */1;
      }
    }
  };
}

function set(h, key, value) {
  var h_buckets = h.buckets;
  var buckets_len = h_buckets.length;
  var i = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_string(0, key)) & (buckets_len - 1 | 0);
  var l = h_buckets[i];
  if (l !== undefined) {
    if (replaceInBucket(key, value, l)) {
      h_buckets[i] = {
        key: key,
        value: value,
        next: l
      };
      h.size = h.size + 1 | 0;
    }
    
  } else {
    h_buckets[i] = {
      key: key,
      value: value,
      next: Bs_internalBucketsType.emptyOpt
    };
    h.size = h.size + 1 | 0;
  }
  if (h.size > (buckets_len << 1)) {
    return resize(h);
  } else {
    return 0;
  }
}

function removeInBucket(h, _, _$1, key, _prec, _buckets) {
  while(true) {
    var buckets = _buckets;
    var prec = _prec;
    if (buckets !== undefined) {
      var cell_next = buckets.next;
      if (buckets.key === key) {
        prec.next = cell_next;
        h.size = h.size - 1 | 0;
        return /* () */0;
      } else {
        _buckets = cell_next;
        _prec = buckets;
        continue ;
        
      }
    } else {
      return /* () */0;
    }
  };
}

function remove(h, key) {
  var h_buckets = h.buckets;
  var i = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_string(0, key)) & (h_buckets.length - 1 | 0);
  var bucket = h_buckets[i];
  if (bucket !== undefined) {
    if (bucket.key === key) {
      h_buckets[i] = bucket.next;
      h.size = h.size - 1 | 0;
      return /* () */0;
    } else {
      return removeInBucket(h, h_buckets, i, key, bucket, bucket.next);
    }
  } else {
    return /* () */0;
  }
}

function getAux(key, _buckets) {
  while(true) {
    var buckets = _buckets;
    if (buckets !== undefined) {
      if (key === buckets.key) {
        return /* Some */[buckets.value];
      } else {
        _buckets = buckets.next;
        continue ;
        
      }
    } else {
      return /* None */0;
    }
  };
}

function get(h, key) {
  var h_buckets = h.buckets;
  var nid = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_string(0, key)) & (h_buckets.length - 1 | 0);
  var match = h_buckets[nid];
  if (match !== undefined) {
    if (key === match.key) {
      return /* Some */[match.value];
    } else {
      var match$1 = match.next;
      if (match$1 !== undefined) {
        if (key === match$1.key) {
          return /* Some */[match$1.value];
        } else {
          var match$2 = match$1.next;
          if (match$2 !== undefined) {
            if (key === match$2.key) {
              return /* Some */[match$2.value];
            } else {
              return getAux(key, match$2.next);
            }
          } else {
            return /* None */0;
          }
        }
      } else {
        return /* None */0;
      }
    }
  } else {
    return /* None */0;
  }
}

function memInBucket(key, _cell) {
  while(true) {
    var cell = _cell;
    if (cell.key === key) {
      return /* true */1;
    } else {
      var match = cell.next;
      if (match !== undefined) {
        _cell = match;
        continue ;
        
      } else {
        return /* false */0;
      }
    }
  };
}

function has(h, key) {
  var h_buckets = h.buckets;
  var nid = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_string(0, key)) & (h_buckets.length - 1 | 0);
  var bucket = h_buckets[nid];
  if (bucket !== undefined) {
    return memInBucket(key, bucket);
  } else {
    return /* false */0;
  }
}

function make(size) {
  return Bs_internalBucketsType.make(/* () */0, /* () */0, size);
}

function size(prim) {
  return prim.size;
}

function ofArray(arr) {
  var len = arr.length;
  var v = Bs_internalBucketsType.make(/* () */0, /* () */0, len);
  for(var i = 0 ,i_finish = len - 1 | 0; i <= i_finish; ++i){
    var match = arr[i];
    set(v, match[0], match[1]);
  }
  return v;
}

function mergeMany(h, arr) {
  var len = arr.length;
  for(var i = 0 ,i_finish = len - 1 | 0; i <= i_finish; ++i){
    var match = arr[i];
    set(h, match[0], match[1]);
  }
  return /* () */0;
}

var N = 0;

var C = 0;

var A = 0;

var clear = Bs_internalBucketsType.clear;

var forEach = Bs_internalBuckets.forEach;

var reduce = Bs_internalBuckets.reduce;

var logStats = Bs_internalBuckets.logStats;

var keepMapInPlace = Bs_internalBuckets.keepMapInPlace;

var toArray = Bs_internalBuckets.toArray;

var copy = Bs_internalBuckets.copy;

var keysToArray = Bs_internalBuckets.keysToArray;

var valuesToArray = Bs_internalBuckets.valuesToArray;

var getBucketHistogram = Bs_internalBuckets.getBucketHistogram;

var isEmpty = Bs_internalBucketsType.isEmpty;

exports.hash = hash;
exports.N = N;
exports.C = C;
exports.A = A;
exports.copyBucketReHash = copyBucketReHash;
exports.resize = resize;
exports.replaceInBucket = replaceInBucket;
exports.set = set;
exports.removeInBucket = removeInBucket;
exports.remove = remove;
exports.getAux = getAux;
exports.get = get;
exports.memInBucket = memInBucket;
exports.has = has;
exports.make = make;
exports.clear = clear;
exports.size = size;
exports.forEach = forEach;
exports.reduce = reduce;
exports.logStats = logStats;
exports.keepMapInPlace = keepMapInPlace;
exports.toArray = toArray;
exports.copy = copy;
exports.keysToArray = keysToArray;
exports.valuesToArray = valuesToArray;
exports.getBucketHistogram = getBucketHistogram;
exports.isEmpty = isEmpty;
exports.ofArray = ofArray;
exports.mergeMany = mergeMany;
/* No side effect */
